src/sentry/incidents/logic.py

```diff
@@ -329,6 +334,30 @@ def bulk_build_incident_query_params(incidents, start=None, end=None, prewindow=
     return query_args_list
 
 
+def calculate_incident_rollup(incident):
+    """
+    This function tries to calculate a good rollup for displaying an incidents
+    stat's on a graph. It prioritized a rollup at the rule's time_window,
+    to be consistent with the builder and more easily understand when/where
+    thresholds are passed. This works for most cases, except for incidents
+    with a long duration and small time window (Results are more than MAX_TIMESERIES_RESULTS - bad)
+    or a short duration and long time window (Very few to no data points - bad).
+    In those cases, we just choose a rollup that results in a reasonable graph.
+    Note: time_window is in minutes, rollup is in seconds.
+    """
+    alert_rule = incident.alert_rule
+    if alert_rule:
+        # TODO: When we persist a rule's time window, switch to using the persisted data.
+        rollup = alert_rule.time_window * 60
+    else:
+        rollup = max(int(incident.duration.total_seconds() / 50), 1)
+
+    while ((incident.duration.total_seconds()) / rollup) > MAX_TIMESERIES_RESULTS:
+        rollup = rollup * 2
```

Comment:
I think this will still end up with weird window sizes. Would it be make sense for us to have a list of sizes and iterate through it? Like 1 min, 10 min, 1 hour, and so on. If we hit the end then we could double until we find something that works.